File: 980. Unique Paths III/main.cpp
#include <vector>

using namespace std;

class Solution {
    int dfs(vector<vector<int>> &grid, int i, int j, int spaces) {
        int rowSize = grid.size();
        int colSize = grid[0].size();

        if (i < 0 || i >= rowSize || j < 0 || j >= colSize || grid[i][j] == -1) {
            return 0;
        }

        if (grid[i][j] == 2) {
            return (spaces == 0) ? 1 : 0;
        }

        int result = 0;

        grid[i][j] = -1;

        result += dfs(grid, i + 1, j, spaces - 1);
        result += dfs(grid, i - 1, j, spaces - 1);
        result += dfs(grid, i, j + 1, spaces - 1);
        result += dfs(grid, i, j - 1, spaces - 1);

        grid[i][j] = 0;

        return result;
    }

public:
    int uniquePathsIII(vector<vector<int>> &grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        int startX = -1;
        int startY = -1;
        int spaces = 1;

        for (int i = 0; i < rowSize; i++) {
            for (int j = 0; j < colSize; j++) {
                if (grid[i][j] == 1) {
                    startX = i;
                    startY = j;
                } else if (grid[i][j] == 0) {
                    spaces++;
                }
            }
        }

        return dfs(grid, startX, startY, spaces);
    }
};

File: 980. Unique Paths III/main.py
from typing import List


class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        def dfs(grid: List[List[int]], i: int, j: int, spaces: int) -> int:
            rowSize = len(grid)
            colSize = len(grid[0])

            if i < 0 or i >= rowSize or j < 0 or j >= colSize or grid[i][j] == -1:
                return 0

            if grid[i][j] == 2:
                return 1 if spaces == 0 else 0

            result = 0

            grid[i][j] = -1

            result += dfs(grid, i + 1, j, spaces - 1)
            result += dfs(grid, i - 1, j, spaces - 1)
            result += dfs(grid, i, j + 1, spaces - 1)
            result += dfs(grid, i, j - 1, spaces - 1)

            grid[i][j] = 0

            return result

        rowSize = len(grid)
        colSize = len(grid[0])
        startX = -1
        startY = -1
        spaces = 1

        for i in range(rowSize):
            for j in range(colSize):
                if grid[i][j] == 1:
                    startX = i
                    startY = j
                elif grid[i][j] == 0:
                    spaces += 1

        return dfs(grid, startX, startY, spaces)
